### PR TITLE
Implement row_number window function

### PR DESCRIPTION
**Pull request by smakalias**
## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by smakalias** [comment]
It seems that peers are not handled in the correct order. I need to investigate more on this.

File: sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java [hunk 1]
```diff
@@ -591,16 +591,19 @@ public Projector visitWindowAgg(WindowAggProjection windowAgg, Context context)
...
-            windowFunctions.add(
-                new AggregateToWindowFunctionAdapter(
-                    ctx.topLevelInputs().toArray(new Input[0]),
-                    (AggregationFunction) impl,
-                    ctx.expressions(),
-                    indexVersionCreated,
-                    bigArrays,
-                    context.ramAccountingContext)
-            );
+            if (impl instanceof AggregationFunction) {
```
Comment (smakalias) [inline sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java]: not sure if I like checking hierarchy vs a type value. 
I can use the `FunctionInfo.Type` instead eg `if (impl.info().type() == FunctionInfo.Type.WINDOW)`.

What is your preference?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java]: If a cast follows it makes imho more sense to use instanceof checks - analysis tooling can then verify that the cast is safe.

**Comment by smakalias** [comment]
> It seems that peers are not handled in the correct order. I need to investigate more on this.

Issue is now fixed

**Comment by smakalias** [comment]
@mfussenegger  pls have another look. The fix for output ordering is also included in this PR as a separate commit

File: sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java [hunk 2]
```diff
@@ -591,16 +591,19 @@ public Projector visitWindowAgg(WindowAggProjection windowAgg, Context context)
...
+                        ctx.topLevelInputs().toArray(new Input[0]),
+                        (AggregationFunction) impl,
+                        ctx.expressions(),
+                        indexVersionCreated,
+                        bigArrays,
+                        context.ramAccountingContext)
+                );
+            } else if (impl instanceof WindowFunction) {
+                windowFunctions.add((WindowFunction) impl);
+            }
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java]: please add a `else` which raises an assertion error that there must be a WindowFunction or an AggregationFunction. It is better to be defensive than to later on have to debug some weird behaviour.

File: sql/src/main/java/io/crate/execution/engine/window/RowNumberWindowFunction.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+import java.util.List;
+
+public class RowNumberWindowFunction {
+
+    private static final String NAME = "row_number";
+
+    static class RowNumberWindowFunctionImplementation implements WindowFunction {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/window/RowNumberWindowFunction.java]: Any reason why we're not having the top-level class implement the `WindowFunction` directly?

Comment (smakalias) [reply sql/src/main/java/io/crate/execution/engine/window/RowNumberWindowFunction.java]: I was thinking of adding later the concrete implementations for `rank()` as `dense_rank()` which share the same signature as `row_number()` and then let the registration in the top-level class (which would be renamed).

What do you think?

Comment (smakalias) [reply sql/src/main/java/io/crate/execution/engine/window/RowNumberWindowFunction.java]: ok, fair comment, I'll change this accordingly.

File: blackbox/docs/general/builtins/window-functions.rst [hunk 4]
```diff
@@ -100,3 +100,23 @@ Example::
    +----------+----------------------------------------+
    SELECT 4 rows in set (... sec)
 
+
+Native Window Functions
```
Comment (andreidan) [inline blackbox/docs/general/builtins/window-functions.rst]: I think we shouldn't use "native" as the concepts of native and "non-native" don't really exist afaik. Aggregates are functions that are optionally treated and executed as window functions (so from an execution perspective they are "native" window functions as well)

Comment (smakalias) [reply blackbox/docs/general/builtins/window-functions.rst]: ok, changing this to "general-purpose" wf

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 5]
```diff
@@ -41,6 +41,7 @@ Breaking Changes
 Changes
 =======
 
+- Added support for native window function ``row_number()``.
```
Comment (andreidan) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: please drop "native" here as well

File: blackbox/docs/general/builtins/window-functions.rst [hunk 6]
```diff
@@ -100,3 +100,23 @@ Example::
...
    SELECT 4 rows in set (... sec)
 
+
+General-Purpose Window Functions
+================================
+
+``row_number()``
+----------------
+
+Returns the number of the current row within its window, counting from 1.
```
Comment (andreidan) [inline blackbox/docs/general/builtins/window-functions.rst]: Not sure we should specify "count from 1" as that's general in SQL

**Pull request merged**